dipy/segment/mrf.pyx

```diff
@@ -0,0 +1,744 @@
+    The standard deviations are set to 1 for all classes.
+
+    Parameters
+    ----------
+    image : array, 
+            3D structural gray-scale image
+    mu : buffer array for the mean of each tissue class
+    sigma : buffer array for the variance of each tissue class
+    
+    Returns
+    -------
+    mu : array, 
+            1 x nclasses, mean of each class
+    sigma : array, 
+            1 x nclasses, standard deviation of each class
+    
+    """
+    cdef:
+        cnp.npy_intp nx = image.shape[0]
+        cnp.npy_intp ny = image.shape[1]
+        cnp.npy_intp nz = image.shape[2]
+        int nclasses = mu.shape[0]
+        int i
+        double min_val
+        double max_val
+    min_val = image[0,0,0]
+    max_val = image[0,0,0]
+    for x in range(nx):
+        for y in range(ny):
+            for z in range(nz):
+                if image[x,y,z] < min_val:
+                    min_val = image[x,y,z]
+                if image[x,y,z] > max_val:
+                    max_val = image[x,y,z]
+    for i in range(nclasses):
+        sigma[i] = 1.0
+        mu[i] = min_val + i * (max_val - min_val)/nclasses
+
+
+cdef void _negloglikelihood(double[:, :, :] image, double[:] mu,
+                            double[:] sigmasq, int classid,
+                            double[:, :, :, :] neglogl) nogil:
+
+    r""" Computes the gaussian negative log-likelihood of each class at 
+    each voxel of `image` assuming a gaussian distribution with means and
+    variances given by `mu` and `sigmasq`, respectively (constant models
+    along the full volume). The negative log-likelihood will be written
+    in `neglogl`.
+    
+    Parameters
+    ----------
+    image : array, 
+            3D structural gray-scale image
+    mu : array, 
+            mean of each class
+    sigmasq : array,
+                variance of each class
+    classid : int,
+                class identifier
+    neglogl : buffer for the neg-loglikelihood
+    
+    Returns
+    -------
+    neglogl : array, 
+                neg-loglikelihood for the class (l = classid)
+
+    """
+
+    cdef:
+        cnp.npy_intp nx = image.shape[0]
+        cnp.npy_intp ny = image.shape[1]
+        cnp.npy_intp nz = image.shape[2]
+        cnp.npy_intp l = classid
+        cnp.npy_intp x, y, z
+        double eps = 1e-8      # We assume images normalized to 0-1
+        double eps_sq = 1e-16  # Maximum precision for double.
+
+    for x in range(nx):
+        for y in range(ny):
+            for z in range(nz):
+                
+                if sigmasq[l] < eps_sq:
+                    
+                    if fabs(image[x, y, z] - mu[l]) < eps:
+                        neglogl[x, y, z, l] = 1 + log(sqrt(2.0 * NPY_PI * sigmasq[l]))
+                    else:
+                        neglogl[x, y, z, l] = NPY_INFINITY
+
+                else:
+                    neglogl[x, y, z, l] = ((image[x, y, z] - mu[l]) ** 2.0) / (2.0 * sigmasq[l])
```

Comment:
I am not sure how to handle line breaks in python. I think I tried breaking it up in a "pythonized" way and it crashed. Any ideas about how to do it correctly? I haven't found anything really explanatory.